python/lsst/daf/butler/registry/versions.py

```diff
@@ -0,0 +1,321 @@
+        return cls(versions)
+
+    @staticmethod
+    def checkCompatibility(old_version: VersionTuple, new_version: VersionTuple, update: bool) -> bool:
+        """Compare two versions for compatibility.
+
+        Parameters
+        ----------
+        old_version : `VersionTuple`
+            Old schema version, typically one stored in a database.
+        new_version : `VersionTuple`
+            New schema version, typically version defined in configuration.
+        update : `bool`
+            If True then read-write access is expected.
+        """
+        if old_version.major != new_version.major:
+            # different major versions are not compatible at all
+            return False
+        if old_version.minor != new_version.minor:
+            # different minor versions are backward compatible for read
+            # access only
+            return new_version.minor > old_version.minor and not update
+        # patch difference does not matter
+        return True
+
+    @staticmethod
+    def schemaDigest(tables: Iterable[sqlalchemy.schema.Table]) -> str:
+        """Calculate digest for a schema.
+
+        Parameters
+        ----------
+        tables : iterable [`sqlalchemy.schema.Table`]
+            Set of tables comprising the schema.
+
+        Returns
+        -------
+        digest : `str`
+            String representation of the digest of the schema.
+
+        Notes
+        -----
+        It is not specified what kind of implementation is used to calculate
+        digest string. The only requrement for that is that result should be
+        stable over time as this digest string will be stored in the
+        configuration and probably in the database too. It should detect (by
+        producing different digests) sensible changes to the schema, but it
+        also should be stable w.r.t. changes that do not actually change the
+        schema (e.g. change in the order of columns or keys.) Current
+        implementation is likely incomplete in that it does not detect all
+        possible changes (e.g. some constraints may not be included into
+        total digest). Digest checking is optional and can be disabled in
+        configuration if configured digest is an empty string, we should delay
+        activating that check until we have a stable implementation for this
+        method.
+        """
+
+        def tableSchemaRepr(table: sqlalchemy.schema.Table) -> str:
+            """Make string representation of a single table schema.
+            """
+            tableSchemaRepr = [table.name]
+            schemaReps = []
+            for column in table.columns:
+                columnRep = f"COL,{column.name},{column.type}"
+                if column.primary_key:
+                    columnRep += ",PK"
+                if column.nullable:
+                    columnRep += ",NULL"
+                schemaReps += [columnRep]
+            for fkConstr in table.foreign_key_constraints:
```

Comment:
Do FK constraints actually affect schema compatibility?  In most of the scenarios I can think of (where there is always a logical FK but we may or may not define an actual constraint for some reason), adding or removing the constraint seems like it'd just be a patch-level change.

Reply:
Even it it is a patch-level change you still probably want to know that your schema is different from a configured version and configured version needs to be updated (together with digest). For me anything that changes schema, even in a trivial way, needs a new version number, so that we can use migration tools to identify those versions.